File: DONGUK/프로그래머스_WEEK 9/3xn 타일링.py
def solution(n):
    answer = 0
    
    if n % 2 == 1:
        return answer
    
    dp = [0] * (n+1)
    
    dp[2] = 3
    dp[4] = 11

    for i in range(6, n+1, 2):
        dp[i] = ((dp[i-2] * 4) - (dp[i-4])) % 1000000007
        
    
    answer = dp[n]
    return answer

File: JUNHEE/프로그래머스_week9/3 x n 타일링.cpp
#include <string>
#include <vector>
#include <iostream>
#define MOD 1000000007
using namespace std;

int solution(int n) {
    int answer = 0;
    long long dp[5001] = {0,};
    dp[0] = 1;
    dp[2] = 3;
    for(int i = 4; i <= n; i += 2){
        dp[i] = dp[i - 2] * 3;
        for(int j = i - 4; j >= 0; j -= 2){
            dp[i] = dp[i] + dp[j] * 2;
        }
        dp[i] = dp[i] % MOD;
    }
    answer = dp[n];
    return answer;
}
